core/templates/cowdata.h

```diff
@@ -293,143 +304,137 @@ void CowData<T>::_unref() {
-		// Wants to clean up.
-		_unref(); // Resets _ptr to nullptr.
-		return OK;
+	const Error error = _fork_allocate(p_size);
+	if (error) {
+		return error;
 	}
 
-	// possibly changing size, copy on write
-	_copy_on_write();
-
-	USize current_alloc_size = _get_alloc_size(current_size);
-	USize alloc_size;
-	ERR_FAIL_COND_V(!_get_alloc_size_checked(p_size, &alloc_size), ERR_OUT_OF_MEMORY);
-
-	if (p_size > current_size) {
-		if (alloc_size != current_alloc_size) {
-			if (current_size == 0) {
-				// alloc from scratch
-				uint8_t *mem_new = (uint8_t *)Memory::alloc_static(alloc_size + DATA_OFFSET, false);
-				ERR_FAIL_NULL_V(mem_new, ERR_OUT_OF_MEMORY);
-
-				SafeNumeric<USize> *_refc_ptr = _get_refcount_ptr(mem_new);
-				USize *_size_ptr = _get_size_ptr(mem_new);
-				T *_data_ptr = _get_data_ptr(mem_new);
-
-				new (_refc_ptr) SafeNumeric<USize>(1); //refcount
-				*(_size_ptr) = 0; //size, currently none
-
-				_ptr = _data_ptr;
-
-			} else {
-				const Error error = _realloc(alloc_size);
-				if (error) {
-					return error;
-				}
-			}
-		}
-
-		// construct the newly created elements
-
+	if (p_size > prev_size) {
 		if constexpr (!std::is_trivially_constructible_v<T>) {
-			for (Size i = *_get_size(); i < p_size; i++) {
+			for (Size i = prev_size; i < p_size; i++) {
 				memnew_placement(&_ptr[i], T);
 			}
 		} else if (p_ensure_zero) {
-			memset((void *)(_ptr + current_size), 0, (p_size - current_size) * sizeof(T));
+			memset((void *)&_ptr[prev_size], 0, (p_size - prev_size) * sizeof(T));
 		}
+	}
 
-		*_get_size() = p_size;
+	return OK;
+}
 
-	} else if (p_size < current_size) {
-		if constexpr (!std::is_trivially_destructible_v<T>) {
-			// deinitialize no longer needed elements
-			for (USize i = p_size; i < *_get_size(); i++) {
-				T *t = &_ptr[i];
-				t->~T();
-			}
-		}
+template <typename T>
+Error CowData<T>::_alloc(USize p_alloc_size) {
+	uint8_t *mem_new = (uint8_t *)Memory::alloc_static(p_alloc_size + DATA_OFFSET, false);
+	ERR_FAIL_NULL_V(mem_new, ERR_OUT_OF_MEMORY);
 
-		if (alloc_size != current_alloc_size) {
-			const Error error = _realloc(alloc_size);
-			if (error) {
-				return error;
-			}
-		}
+	_ptr = _get_data_ptr(mem_new);
 
-		*_get_size() = p_size;
-	}
+	// If we alloc, we're guaranteed to be the only reference.
+	new (_get_refcount()) SafeNumeric<USize>(1);
 
 	return OK;
 }
 
 template <typename T>
-Error CowData<T>::_realloc(Size p_alloc_size) {
+Error CowData<T>::_realloc(USize p_alloc_size) {
 	uint8_t *mem_new = (uint8_t *)Memory::realloc_static(((uint8_t *)_ptr) - DATA_OFFSET, p_alloc_size + DATA_OFFSET, false);
 	ERR_FAIL_NULL_V(mem_new, ERR_OUT_OF_MEMORY);
 
-	SafeNumeric<USize> *_refc_ptr = _get_refcount_ptr(mem_new);
-	T *_data_ptr = _get_data_ptr(mem_new);
+	_ptr = _get_data_ptr(mem_new);
 
 	// If we realloc, we're guaranteed to be the only reference.
-	new (_refc_ptr) SafeNumeric<USize>(1);
-	_ptr = _data_ptr;
+	// So the reference was 1 and was copied to be 1 again.
```

Comment:
could we add an assert for that for debug builds ?

Reply:
Good idea!

Reply:
Seems like there's nothing like `DEBUG_ASSERT`, so I think `DEV_ASSERT` will be good enough?

Reply:
probably yes